### PR TITLE
[Doc] Fix broken references in data documentation

### PR DESCRIPTION
**Pull request by peytondmurray**
## Why are these changes needed?

This PR fixes some bad references in Ray Data.

## Related issue number

Partially addresses #39658.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/datasource.py [hunk 1]
```diff
@@ -17,16 +17,16 @@ class Datasource:
     @Deprecated
     def create_reader(self, **read_args) -> "Reader":
         """
-        Deprecated: Implement :meth:`Datasource.get_read_tasks` and
-        :meth:`Datasource.estimate_inmemory_data_size` instead.
+        Deprecated: Implement :meth:`~ray.data.Datasource.get_read_tasks` and
+        :meth:`~ray.data.Datasource.estimate_inmemory_data_size` instead.
```
Comment (bveeramani) [inline python/ray/data/datasource/datasource.py:21]: For my own understanding, do I always need to qualify the class name, even if the class is defined in the same module?

File: python/ray/data/iterator.py [hunk 2]
```diff
@@ -271,7 +271,7 @@ def iter_torch_batches(
 
         This iterable yields a dictionary of column-tensors. If you are looking for
         more flexibility in the tensor conversion (e.g. casting dtypes) or the batch
-        format, try using :meth:`~ray.data.iterator.DataIterator.iter_batches` directly.
+        format, try using :py:meth:`~ray.data.DataIterator.iter_batches` directly.
```
Comment (bveeramani) [inline python/ray/data/iterator.py:274]: Why is the `:py:` needed here?

**Review by bveeramani — APPROVED** [review]

File: python/ray/data/datasource/datasource.py (hunk 1 above)
Comment (peytondmurray) [reply python/ray/data/datasource/datasource.py:21]: It depends on how you have Sphinx configured, and the syntax you use. I feel like the simplest thing to do is fully qualify it though.

File: python/ray/data/iterator.py (hunk 2 above)
Comment (peytondmurray) [reply python/ray/data/iterator.py:274]: Probably not necessary here - in this case it's equivalent to `:meth:`. I was messing around trying to get this to build correctly and must have put it in before realizing I needed to remove `.iterator`. I can remove it.

Comment (peytondmurray) [reply python/ray/data/iterator.py]: Removed!

**Review by aslonnie — APPROVED** [review]

**Pull request merged**